onert-micro/onert-micro/src/train/train_optimizers/Adam.cpp

```diff
@@ -0,0 +1,263 @@
+  {
+    // If not - let's just move it with calculations
+    // Goes over all calculated gradients
+    // Warning: assume that backward storage at this moment contains only weights gradients -
+    // This should be done due to execution plan work
+    for (auto &tensor_to_data : backward_tensor_to_data)
+    {
+      // Move data
+      _tensor_index_to_gradient[tensor_to_data.first] = tensor_to_data.second;
+      tensor_to_data.second = nullptr;
+    }
+    backward_tensor_to_data.clear();
+  }
+  else
+  {
+    // Goes over all calculated gradients
+    // Warning: assume that backward storage at this moment contains only weighs gradients -
+    // This should be done due to execution plan work
+    for (auto &tensor_to_data : backward_tensor_to_data)
+    {
+      auto tensor = context.getTensorByIndex(tensor_to_data.first);
+      core::OMRuntimeShape shape(tensor);
+
+      const auto flat_size = shape.flatSize();
+
+      auto *grad_data = reinterpret_cast<float *>(_tensor_index_to_gradient[tensor_to_data.first]);
+      auto *calculated_data = reinterpret_cast<float *>(tensor_to_data.second);
+
+      for (uint32_t i = 0; i < flat_size; ++i)
+      {
+        grad_data[i] += calculated_data[i];
+      }
+    }
+  }
+
+  return Ok;
+}
+
+/*
+ * Update internal states according to calculated gradients using Adam theory
+ * m(t) = beta_1 * m(t-1) + (1 - beta_1) * calculated_gradients(t)
+ * v(t) = beta_2 * v(t-1) + (1 - beta_2) * (calculated_gradients(t)) ^ 2
+
+ * Update weights according to Adam theory:
+ * m`(t) = m(t) / (1 - (beta_1) ^ t)
+ * v`(t) = v(t) / (1 - (beta_2) ^ t)
+ *
+ * w(t + 1) = w(t) - lambda * m`(t) / (sqrt(v` + epsilon))
+ */
+OMStatus Adam::updateWeights(const onert_micro::OMTrainingContext &training_config,
+                             core::OMRuntimeContext &context)
+{
+  assert(!_tensor_index_to_gradient.empty());
+
+  for (auto &tensor_to_data : _tensor_index_to_gradient)
+  {
+    auto exponent_squares_it = _tensor_to_exponent_avg_squares.find(tensor_to_data.first);
+    if (exponent_squares_it == _tensor_to_exponent_avg_squares.end())
+      return UnknownError;
+
+    auto exponent_it = _tensor_to_exponent_avg.find(tensor_to_data.first);
+    if (exponent_it == _tensor_to_exponent_avg.end())
+      return UnknownError;
+
+    auto tensor = context.getTensorByIndex(tensor_to_data.first);
+    core::OMRuntimeShape shape(tensor);
+
+    const auto flat_size = shape.flatSize();
+
+    auto *exponent_data = reinterpret_cast<float *>(exponent_it->second);
+    auto *exponent_square_data = reinterpret_cast<float *>(exponent_squares_it->second);
+    auto *calculated_data = reinterpret_cast<float *>(tensor_to_data.second);
+    float beta = training_config.beta;
+    float beta_squares = training_config.beta_squares;
+    auto batches = static_cast<float>(training_config.batch_size);
+    for (uint32_t i = 0; i < flat_size; ++i)
+    {
+      const auto cur_val = calculated_data[i] / batches;
```

Comment:
```suggestion
      const auto cur_val = calculated_data[i];
```
not sure but, this one is correct ?